File: backend/models/achievement_version_record.py
from sqlalchemy import Column, String, DateTime, Text, Float, JSON
from sqlalchemy.orm import relationship
from app.db.database import Base


class AchievementVersionRecord(Base):
    """成果变更记录表 - 专门记录所有成果的版本变更历史"""
    __tablename__ = "achievement_version_records"

    id = Column(String(50), primary_key=True, index=True, comment="变更记录ID")
    achievement_name = Column(String(100), nullable=False, index=True, comment="成果名称")
    achievement_id = Column(String(50), nullable=False, index=True, comment="成果ID（基于achievement_name + product_external_version）")
    product_external_version = Column(String(20), nullable=False, index=True, comment="产品对外版本号")
    from_version = Column(String(20), nullable=False, comment="变更前版本号")
    to_version = Column(String(20), nullable=False, comment="变更后版本号")
    changed_fields = Column(JSON, nullable=False, comment="变更字段（JSON格式，记录变更前后的字段值差异）")
    change_description = Column(Text, comment="变更说明")
    risk_tags = Column(JSON, comment="关联的风险标签ID（JSON数组格式）")
    operator = Column(String(50), comment="操作人")
    change_time = Column(DateTime(timezone=True), nullable=False, comment="变更时间")
    created_at = Column(DateTime(timezone=True), nullable=False, comment="创建时间")

    def __repr__(self):
        return f"<AchievementVersionRecord(id={self.id}, achievement_name={self.achievement_name}, from_version={self.from_version}, to_version={self.to_version})>"

File: java-backend/src/main/java/com/example/achievement/entity/AchievementVersionRecord.java
package com.example.achievement.entity;

import javax.persistence.*;
import lombok.Data;

import java.time.LocalDateTime;
import java.util.List;

@Entity
@Table(name = "achievement_version_records")
@Data
public class AchievementVersionRecord {

    @Id
    @Column(length = 50)
    private String id;

    @Column(name = "achievement_name", nullable = false, length = 100)
    private String achievementName;

    @Column(name = "achievement_id", nullable = false, length = 50)
    private String achievementId;

    @Column(name = "product_external_version", nullable = false, length = 20)
    private String productExternalVersion;

    @Column(name = "from_version", nullable = false, length = 20)
    private String fromVersion;

    @Column(name = "to_version", nullable = false, length = 20)
    private String toVersion;

    @Column(name = "changed_fields", nullable = false, columnDefinition = "TEXT")
    @org.hibernate.annotations.Type(type = "text")
    private String changedFields;

    @Column(name = "change_description", columnDefinition = "TEXT")
    private String changeDescription;

    @Column(name = "risk_tags", columnDefinition = "TEXT")
    @org.hibernate.annotations.Type(type = "text")
    private String riskTags;

    @Column(length = 50)
    private String operator;

    @Column(name = "change_time", nullable = false)
    private LocalDateTime changeTime;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        if (changeTime == null) {
            changeTime = LocalDateTime.now();
        }
    }
}
